File: LeetCode/LeetCode-Java/2020JuneLeetCodingChallenge/Insert Delete GetRandom O(1).java
class RandomizedSet {
    // 感觉java的操作时间复杂度不如C++的明确
    ArrayList<Integer> elements;
    HashMap<Integer,Integer> hash_table;
    java.util.Random rand;
    
    /** Initialize your data structure here. */
    public RandomizedSet() {
        elements = new ArrayList<Integer>();
        hash_table = new HashMap<Integer,Integer>();
        rand = new java.util.Random(42);
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        if(hash_table.containsKey(val)) { // 如果存在val，则直接返回false
            return false;
        }
        // hash_table中记录val的值和val对应在数组中的索引位置
        hash_table.put(val, elements.size());
        elements.add(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        if(!hash_table.containsKey(val)) { // 如果不包含val，则直接返回false
            return false;
        }
        int loc = hash_table.get(val); // 获取val在数组中对应的索引
        if(loc < elements.size() - 1) { // 如果此时删除的val不在最后一位，则把最后一位与val进行交换
            int lastone = elements.get(elements.size() - 1); // 获取最后一位的数据
            elements.set(loc, lastone); // 要被删除的val对应的位置上设置为最后一位的数据，最后一位可以不用设置为val，因为要被删除了
            hash_table.put(lastone, loc); // 同时修改table中最后一位对应的索引位置
            
        }
        hash_table.remove(val);
        elements.remove(elements.size() - 1); // 删除该元素，当删除的位置是第一个或者最后一个时是O(1)
        return true;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
        return elements.get(rand.nextInt(elements.size()));
    }
    
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: LeetCode/LeetCode-CPP/2020JuneLeetCodingChallenge/Insert Delete GetRandom O(1).cpp
class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        n = 0;
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(hash_table[val] != 0) {
            return false;
        }
        // 第一次添加该数，加1
        n++;
        hash_table[val] = n; // val对应的值是val在数组中的索引+1
        if(elements.size() >= n) { // 需要注意数组空间的位置关系，确保hashtable中的val对应在数组中的位置为n
            elements[n-1] = val;
        }
        elements.push_back(val); // 放入数组中用来查看。
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        // 如果==0，则说明此时不存在val
        if(hash_table[val] == 0) {
            return false;
        }
        int loc = hash_table[val]; // 找到val对应数组中的索引
        // loc < n 时说明此时删除的元素不在数组的最后一位，需要调整最后以为与当前删除的位置
        if(loc < n) { // 相当于把要删除的元素变成数组中的最后一个
            int lastone = elements[n - 1]; // 找到最后一位元素
            elements[loc - 1] = lastone; // 把要删除val的位置设置为最后一位元素
            hash_table[lastone] = loc; // 对应的hashtable中的lastone位置也要进行修改
        }
        hash_table[val] = 0; // hashtable中删除
        elements[n - 1] = 0; // 数组中删除（标记为0）
        n--;
        // cout << "delete after" << val << ":";
        // for(auto n : elements) {
        //     cout << n << " ";
        // }
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return elements[rand() % n];
    }
private:
    unordered_map<int,int> hash_table; // 用hashtable来检测是否存在，以及获取数据
    vector<int> elements; // 用来存放元素，可以根据随机数来直接获取数据
    int n; // 用来标识有多少个元素
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */